milhoja/core.py

```diff
@@ -71,97 +67,64 @@ def __enter__(self):
-        note = self.repo.lookup_note(commit.id.__str__(), __notes_template_ref__)
-
-        info = json.loads(note.message)
-        assert isinstance(info[__key_source__], six.string_types)
-        assert isinstance(info[__key_reference__], six.string_types)
-
-        return info[__key_source__], info[__key_reference__]
-
-    def get_context(self):
-        commit = self.get_last_commit()
-        note = self.repo.lookup_note(commit.id.__str__(), __notes_template_ref__)
-
-        info = json.loads(note.message)
-        assert isinstance(info[__key_context__], dict)
-
-        return info[__key_context__]
-
-    def create_note(self, commit, source, reference, context):
-        # Create Git Note with serialized template references and context
-        self.repo.create_note(
-            json.dumps({
-                __key_source__: source,
-                __key_reference__: reference,
-                __key_context__: context
-            }),
-            self.repo.default_signature,
-            self.repo.default_signature,
-            commit.hex,
-            __notes_template_ref__
-        )
-
-    def merge_template_branch(self, message):
+    def merge_template_branch(self, message: str):
         # Lookup template branch
-        branch = self.repo.lookup_branch(__template_branch__)
+        branch = self.repo.lookup_branch(TEMPLATE_BRANCH)
 
         # Analyze merge between template branch and HEAD
         analysis, _ = self.repo.merge_analysis(branch.target)
 
-        # Up to date, nothing to do
         if analysis & GIT_MERGE_ANALYSIS_UP_TO_DATE:
-            # Should append if user reinstalled an existing template
-            # TODO log this info to user
-            pass
+            logger.info('The branch is already up to date, no need to merge.')
 
         elif analysis & GIT_MERGE_ANALYSIS_FASTFORWARD or analysis & GIT_MERGE_ANALYSIS_NORMAL:
-            # Let's merge template changes.
+            # Let's merge template changes using --allow-unrelated-histories. This will allow 
+            # the disjoint histories to be merged successfully. If you want to manually replicate
+            # this option please run:
+            #
+            #     "git merge --allow-unrelated-histories template"
+            #
             self.repo.merge(branch.target)
 
-            # TODO What to do with conflict ?
-            #        -> Let user resolve ?
-            #        -> Raise an error after analysis ?
-            assert self.repo.index.conflicts is None
+            # If there is a conflict we should error and let the user manually resolve it.
+            if self.repo.index.conflicts is not None:
+                raise MergeConflictException(
+                    f'Cannot merge the template commit ({branch.target}) with the current HEAD '
+                    f'({self.repo.head}). Please resolve them manually and run \'git commit\' '
+                    'to merge'
+                )
```

Comment:
Does it make sense to raise an exception here? For 99% of the cases, we would expect a merge conflict. That doesn't seem like an exceptional condition to me. Maybe this is handled elsewhere though.

Reply:
I'm raising an exception in lieu of anything better to do at this point as we cannot automatically proceed unfortunately. Maybe we can handle and display this message a little nicer at the `cli.py` layer but to me it still seems like an exception even if it's going to be the default behavior essentially.

Reply:
Cool after a discussion offline I decided to create https://zbrt.atl.zillow.net/browse/ZOMLINFRA-427 to handle this exception at the CLI layer more nicely and give the user instructions on the best path forward in this case.